### PR TITLE
MBS-6381: Fetch images for entities from Wikidata

### PR DESCRIPTION
**Pull request by gentlecat**
For example, to get image for label [bf519cd2-4e1e-46bf-948b-615f75424606](https://beta.musicbrainz.org/label/bf519cd2-4e1e-46bf-948b-615f75424606) we can use https://www.wikidata.org/w/api.php?action=wbgetclaims&format=json&entity=Q330629&property=P154.

*http://tickets.musicbrainz.org/browse/MBS-6381*

### CODE REVIEW

File: lib/MusicBrainz/Server/Data/WikidataProperty.pm [hunk 1]
```diff
@@ -0,0 +1,54 @@
+package MusicBrainz::Server::Data::WikidataProperty;
+use Moose;
+use namespace::autoclean;
+
+use Readonly;
+
+with 'MusicBrainz::Server::Data::Role::MediaWikiAPI';
+
+Readonly my $WIKIDATA_CACHE_TIMEOUT => 60 * 60 * 24 * 3; # 3 days
```
Comment (gentlecat) [inline lib/MusicBrainz/Server/Data/WikidataProperty.pm]: Not sure if this timeout is any good in this case.

Comment (mwiencek) [inline lib/MusicBrainz/Server/Data/WikidataProperty.pm]: Seems pretty reasonable.

**Comment by gentlecat** [comment]
So now it checks if there a *picture* (Wikimedia Commons) relationship first. If there's no picture, it checks associated Wikidata entity for a *logo image* property ([P154](https://www.wikidata.org/wiki/Property:P154)).

The only additional entity type I enabled this for is Label. Are there any other that could use a logo image?

I also need to figure out how to use *image* ([P18](https://www.wikidata.org/wiki/Property:P18)) and *locator map image* ([P242](https://www.wikidata.org/wiki/Property:P242)) properties.

**Comment by mwiencek** [comment]
I thought series might have logos but I can't seem to find any on Wikidata. I found several with images though (P18).

**Comment by gentlecat** [comment]
OK, should be ready now. Let me know if I need to add or fix something.

File: lib/MusicBrainz/Server/Controller/Role/CommonsImage.pm [hunk 2]
```diff
@@ -21,24 +25,53 @@ sub commons_image : Chained('load') PathPart('commons-image')
...
+
+    my ($wikidata_link) = map {
+            $_->target;
+        } grep {
+            $_->target->isa('MusicBrainz::Server::Entity::URL::Wikidata')
+        } @{ $entity->relationships_by_link_type_names('wikidata') };
+
+    if ($wikidata_link) {
+        my $properties = $c->model('WikidataProperties')->get_wikidata_properties($wikidata_link->pretty_name);
+        if ($entity->isa('MusicBrainz::Server::Entity::Area') &&
```
Comment (gentlecat) [inline lib/MusicBrainz/Server/Controller/Role/CommonsImage.pm]: Should we keep this constraint?

File: lib/MusicBrainz/Server/Data/WikidataProperties.pm [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+with 'MusicBrainz::Server::Data::Role::MediaWikiAPI';
+
+Readonly my $WIKIDATA_CACHE_TIMEOUT => 60 * 60 * 24 * 3; # 3 days
+
+sub get_wikidata_properties {
+    my ($self, $entity, $property) = @_;
+
+    my $url_pattern = "https://www.wikidata.org/w/api.php?action=wbgetclaims&format=json&entity=%s%s";
+    return $self->_fetch_cache_or_url($url_pattern,
+                                      "wikidata_property",
```
Comment (gentlecat) [inline lib/MusicBrainz/Server/Data/WikidataProperties.pm]: This is basically a cache key. Might be a bit confusing due to how `MediaWikiAPI` is written.

**Comment by mwiencek** [comment]
I think we should probably add Role::CommonsImage to all the entity types that have wikidata relationships (besides releases and release groups, which are showing cover art). That leaves Event, Series, and Work.

File: lib/MusicBrainz/Server/Controller/Role/CommonsImage.pm (hunk 2 above)
Comment (mwiencek) [inline lib/MusicBrainz/Server/Controller/Role/CommonsImage.pm]: Well, dunno if any of the other entity types would even have P242, but seems okay to only show those for areas.

**Comment by gentlecat** [comment]
Yeah, makes sense. Done. :ok_hand:

File: lib/MusicBrainz/Server/Controller/Role/CommonsImage.pm (hunk 2 above)
Comment (gentlecat) [inline lib/MusicBrainz/Server/Controller/Role/CommonsImage.pm]: OK

**Comment by mwiencek** [comment]
Works great, very cool :shipit:

**Pull request merged**